presto-main/src/main/java/com/facebook/presto/operator/unnest/Unnester.java

```diff
@@ -14,32 +14,32 @@
 package com.facebook.presto.operator.unnest;
 
 import com.facebook.presto.spi.block.Block;
-import com.facebook.presto.spi.block.PageBuilderStatus;
 import com.facebook.presto.spi.type.Type;
+import org.openjdk.jol.info.ClassLayout;
 
-import static com.google.common.base.Preconditions.checkState;
+import static com.facebook.presto.array.Arrays.ensureCapacity;
+import static io.airlift.slice.SizeOf.sizeOf;
 import static java.util.Objects.requireNonNull;
 
 /**
  * Generic Unnester implementation for nested columns.
- *
+ * <p>
  * This is a layer of abstraction between {@link UnnestOperator} and {@link UnnestBlockBuilder} to enable
  * translation of indices from input nested blocks to underlying element blocks.
  */
 abstract class Unnester
 {
-    private final UnnestBlockBuilder[] unnestBlockBuilders;
-    private int currentPosition;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(Unnester.class).instanceSize();
+
+    protected UnnestBlockBuilder[] unnestBlockBuilders;
+    protected int currentPosition;
+    protected int[] entryCounts = new int[0];
 
     protected Unnester(Type... types)
```

Comment:
this constructor doesn't use the types; it just needs the total count; hence, let's change it to

```
    protected Unnester(int fieldCount)
    {
        unnestBlockBuilders = new UnnestBlockBuilder[fieldCount];
        currentPosition = 0;
    }
```

Then, change the subclasses to avoid re-allocating unnestBlockBuilders and re-initializing currentPosition:

```
    public ArrayOfRowsUnnester(Type... types)
    {
        super(types.length);
        this.fieldCount = types.length;
        for (int i = 0; i < fieldCount; i++) {
            unnestBlockBuilders[i] = new ArrayOfRowUnnestBlockBuilder(types[i], entryCounts);
        }
    }

    public MapUnnester(Type keyType, Type valueType)
    {
        super(2);
        unnestBlockBuilders[0] = new GenericUnnestBlockBuilder(keyType, entryCounts);
        unnestBlockBuilders[1] = new GenericUnnestBlockBuilder(valueType, entryCounts);
    }

    public ArrayUnnester(Type elementType)
    {
        super(1);
        unnestBlockBuilders[0] = new GenericUnnestBlockBuilder(elementType, entryCounts);
    }
```